### PR TITLE
Move color palette editor to asset editor

### PR DESCRIPTION
**Pull request by kimprice**
This is the initial move of the color palette editor from experiments to the asset editor. (Refer to https://github.com/microsoft/pxt-arcade/issues/5327 and https://github.com/microsoft/pxt/pull/9000) The functionality is the same, with minor UI changes.

TO DO: 
- update UI according to new designs,
- add functionality to save/delete custom palettes. Currently, you can have one custom palette, but if you select a built-in palette from the dropdown, you will lose the custom palette.
- make updating the project smoother. Right now, after closing the palette editor modal, it flashes to the blocks view and then goes back to the asset editor view.

Try it out here: https://arcade.makecode.com/app/575d5d9de74b299a768c444de8f27ffb9b964d05-fa87e8108b

image
image
image

### CODE REVIEW

**Comment by abchatra** [comment]
Clicking on the color palette in the build uploaded does nothing for me. Does it work for you

**Comment by kimprice** [comment]
> Clicking on the color palette in the build uploaded does nothing for me. Does it work for you

Yes, it works for me. Are you getting any console errors?

**Comment by srietkerk** [comment]
This is happening for me as well. 

A gif of what's happening: 
color-palette


Errors that I'm getting: 
palette-errors

**Comment by jwunderl** [comment]
Yeah, I'm getting this error during a react crash: 
```
Uncaught TypeError: Cannot read properties of undefined (reading 'slice')
    at PaletteSwatch (PaletteSwatch.js:7:1)
    at renderWithHooks (react-dom.development.js:14985:1)
    at mountIndeterminateComponent (react-dom.development.js:17811:1)
    at beginWork (react-dom.development.js:19049:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
    at invokeGuardedCallback (react-dom.development.js:4056:1)
    at beginWork$1 (react-dom.development.js:23964:1)
    at performUnitOfWork (react-dom.development.js:22776:1)
    at workLoopSync (react-dom.development.js:22707:1)
```

**Comment by kimprice** [comment]
Thanks to all who tested the first link. I fixed the bug and updated the link in the description. 

I was grabbing the initial palette info from the pxt.json file which Joey informed me would have an undefined palette for new projects. The link was working for me because the palette was already defined in my project. I updated it to take the palette info from runtime if the palette in the pxt.json file is undefined.

File: react-common/components/palette/ColorPickerField.tsx [hunk 1]
```diff
@@ -42,24 +21,30 @@ export const ColorPickerField = (props: ColorPickerFieldProps) => {
...
+        if (newValue?.[0] != '#') {
+            newValue = "#" + newValue;
+        }
+        if (newValue.length > 7) {
+            newValue = newValue.substring(0,7);
+        }
         if (/#[0-9a-fA-F]{6}/.test(newValue)) {
-            onColorChanged(newValue);
+            onColorChanged(newValue.toUpperCase());
         }
```
Comment (srietkerk) [inline react-common/components/palette/ColorPickerField.tsx:36]: I experience some weirdness with this functionality on the first color in the list. If I try to enter any text, it doesn't acknowledge it. This works fine with the other colors in the list. 

https://user-images.githubusercontent.com/49178322/206545046-60389988-0a72-463b-9389-e2ad4ab05c97.mp4

Comment (srietkerk) [reply react-common/components/palette/ColorPickerField.tsx:36]: I'm also a bit worried about an edge case where users just spam the keyboard to see what happens in the text field and the color that was there is lost. 

However, I also think that if users are wanting to experiment like that, then it's fair for that progress to be lost. This might be something that we just want to see if users get upset about and change the logic later.

Just wanted to comment as something we might want to keep in mind.

File: react-common/components/palette/ColorPickerField.tsx [hunk 2]
```diff
@@ -42,24 +21,30 @@ export const ColorPickerField = (props: ColorPickerFieldProps) => {
...
-        setCurrentColor(e.target.value);
+        setCurrentColor(e.target.value.toUpperCase());
     }
 
     const onTextInputChanged = (newValue: string) => {
+        if (newValue?.[0] != '#') {
+            newValue = "#" + newValue;
+        }
+        if (newValue.length > 7) {
+            newValue = newValue.substring(0,7);
```
Comment (jwunderl) [inline react-common/components/palette/ColorPickerField.tsx:32]:
```suggestion
            newValue = newValue.substring(0, 7);
```

nit

File: react-common/components/palette/PaletteEditor.tsx [hunk 3]
```diff
@@ -45,22 +53,15 @@ export const PaletteEditor = (props: PaletteEditorProps) => {
...
-                    color={c}
-                    onColorChanged={newColor => updateColor(i, newColor)}
-                    onMoveColor={up => moveColor(i, up)}
-                    disabled={i === 0}
-                />
-            )}
-        </div>
-    </Modal>
+    return <div className="common-palette-editor">
+        {(currentPalette).colors.slice(1).map((c, i) =>
```
Comment (jwunderl) [inline react-common/components/palette/PaletteEditor.tsx:57]:
```suggestion
        {currentPalette.colors.slice(1).map((c, i) =>
```

nit

File: webapp/src/components/assetEditor/assetSidebar.tsx [hunk 4]
```diff
@@ -189,35 +201,42 @@ class AssetSidebarImpl extends React.Component<AssetSidebarProps, AssetSidebarSt
...
-                    onClick={this.copyAssetHandler} />}
-                {canDelete && <Button
-                    label={lf("Delete Asset")}
-                    title={lf("Delete the selected asset")}
-                    ariaLabel={lf("Delete the selected asset")}
-                    className="delete-asset"
-                    onClick={this.showDeleteModal} />}
-            </List>}
+            <List className="asset-editor-sidebar-controls">
+                    { asset && canEdit && <Button
```
Comment (jwunderl) [inline webapp/src/components/assetEditor/assetSidebar.tsx:205]:
```suggestion
                { asset && canEdit && <Button
```
nit: this section is double indented until the asset palette button; unindenting will also remove a lot of the diff which is nice :)

File: react-common/components/palette/ColorPickerField.tsx (hunk 1 above)
Comment (kimprice) [reply react-common/components/palette/ColorPickerField.tsx:36]: I'm having trouble reproducing the bug with the first color. @jwunderl was this happening for you as well?

> I'm also a bit worried about an edge case where users just spam the keyboard to see what happens in the text field and the color that was there is lost.

Yea, ideally the text input wouldn't let you write any non-hex characters. I'm not sure why currently the text validation doesn't happen until on blur. But if the value is invalid, it should change back to the previous value.

File: react-common/components/palette/PaletteEditor.tsx [hunk 5]
```diff
@@ -11,10 +11,18 @@ export interface PaletteEditorProps {
 export const PaletteEditor = (props: PaletteEditorProps) => {
     const { palette, onPaletteChanged } = props;
 
-    const [currentPalette, setCurrentPalette] = React.useState<Palette | undefined>(undefined);
+    const [currentPalette, setCurrentPalette] = React.useState<Palette | undefined>(palette);
```
[the remark this answers is not in the document]
Comment (kimprice) [reply react-common/components/palette/PaletteEditor.tsx:14]: I'm not sure if one way is better than the other. I am pulling in useEffect, so I could do that for useState too. This `React.useState` is just the way it was written originally. @jwunderl?

[the remark this answers is not in the document]
Comment (jwunderl) [reply react-common/components/palette/PaletteEditor.tsx:14]: It really doesn't matter. You *might* lean towards importing the same way as useEffect just for the slightest consistency, but there is zero performance / behavioral difference

File: react-common/components/palette/PaletteEditor.tsx [hunk 6]
```diff
@@ -45,22 +53,15 @@ export const PaletteEditor = (props: PaletteEditorProps) => {
...
-        </div>
-    </Modal>
+    return <div className="common-palette-editor">
+        {(currentPalette).colors.slice(1).map((c, i) =>
+            <ColorPickerField
+                key={i}
+                index={i}
+                color={c}
+                onColorChanged={newColor => updateColor(i + 1, newColor)}
+                onMoveColor={up => moveColor(i + 1, up)}
```
[the remark this answers is not in the document]
Comment (kimprice) [reply react-common/components/palette/PaletteEditor.tsx:63]: I'm not able to repro this. Is anyone else able to?

[the remark this answers is not in the document]
Comment (srietkerk) [reply react-common/components/palette/PaletteEditor.tsx:63]: Writing here for transparency Not sure how I got this the first time, but I re-visited the link and haven't gotten this issue again.

File: webapp/src/components/assetEditor/assetSidebar.tsx (hunk 4 above)
Comment (kimprice) [reply webapp/src/components/assetEditor/assetSidebar.tsx]: nice catch! thanks

File: webapp/src/components/assetEditor/assetSidebar.tsx [hunk 7]
```diff
@@ -189,35 +201,42 @@ class AssetSidebarImpl extends React.Component<AssetSidebarProps, AssetSidebarSt
...
+                    onClick={this.duplicateAssetHandler} />}
+                {asset && canCopy && <Button
                     label={lf("Copy")}
                     title={lf("Copy the selected asset to the clipboard")}
                     ariaLabel={lf("Copy the selected asset to the clipboard")}
                     leftIcon="icon paste"
                     className="asset-editor-button"
                     onClick={this.copyAssetHandler} />}
-                {canDelete && <Button
-                    label={lf("Delete Asset")}
```
Comment (riknoll) [inline webapp/src/components/assetEditor/assetSidebar.tsx:215]: seems like this change is unrelated to this pr

File: react-common/styles/palette/PaletteSwatch.less [hunk 8]
```diff
@@ -11,6 +11,9 @@
 .common-palette-color {
     width: 1rem;
     height: 1rem;
+    border-color: black;
+    border-width: thin;
+    border-style: solid;
```
Comment (riknoll) [inline react-common/styles/palette/PaletteSwatch.less:16]:
```suggestion
    border: 1px solid #000;
```

**Review by riknoll — APPROVED** [review]

**Comment by riknoll** [comment]
Maybe don't reload if the palette didn't change?

File: webapp/src/components/assetEditor/assetSidebar.tsx (hunk 7 above)
Comment (kimprice) [reply webapp/src/components/assetEditor/assetSidebar.tsx:215]: Yea, it's slightly unrelated. We felt the delete button needed to be more consistent with the other asset action buttons in order to better distinguish them from the more high-level palette button. Also aesthetically, it looked odd to have a button with no background in the middle of the set of buttons.

**Comment by abchatra** [comment]
🏆
